core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,302 @@
+    this.immm = match.getModule(ItemModifyMatchModule.class);
+
+    fillers.forEach(
+        filler -> {
+          if (filler.getRefillTrigger() == null) return;
+          match
+              .needModule(FilterMatchModule.class)
+              .onRise(
+                  Match.class,
+                  filler.getRefillTrigger(),
+                  m -> {
+                    this.filledAt.keySet().removeIf(f -> filler.equals(f.getRight()));
+                  });
+        });
+  }
+
+  /**
+   * Return a predicate that applies a Filter to the given InventoryHolder, or null if the
+   * InventoryHolder is not something that we should be filling.
+   */
+  private static @Nullable Predicate<Filter> filterPredicate(InventoryHolder holder) {
+    if (holder instanceof DoubleChest) {
+      final DoubleChest doubleChest = (DoubleChest) holder;
+      return filter ->
+          !filter.query(new BlockQuery((Chest) doubleChest.getLeftSide())).isDenied()
+              || !filter.query(new BlockQuery((Chest) doubleChest.getRightSide())).isDenied();
+    } else if (holder instanceof BlockState) {
+      return filter -> !filter.query(new BlockQuery((BlockState) holder)).isDenied();
+    } else if (holder instanceof Player) {
+      // This happens with crafting inventories, and possibly other transient inventory types
+      // Pretty sure we never want to fill an inventory held by the player
+      return null;
+    } else if (holder instanceof Entity) {
+      return filter -> !filter.query(new EntityQuery((Entity) holder)).isDenied();
+    } else {
+      // If we're not sure what it is, don't fill it
+      return null;
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void onInventoryOpen(InventoryOpenEvent event) {
+    final MatchPlayer opener = this.match.getParticipant(event.getActor());
+    if (opener == null) return;
+
+    final Inventory inventory = event.getInventory();
+    final Predicate<Filter> filterPredicate = filterPredicate(inventory.getHolder());
+    if (filterPredicate == null) return;
+
+    logger.fine(
+        () -> opener.getName() + " opened a " + inventory.getHolder().getClass().getSimpleName());
+
+    // Find all Fillers that apply to the holder of the opened inventory
+    final List<FillerDefinition> fillers =
+        this.fillers.stream()
+            .filter(filler -> filterPredicate.test(filler.getFilter()))
+            .collect(Collectors.toList());
+    if (fillers.isEmpty()) return;
+
+    logger.fine(
+        () ->
+            "Found fillers "
+                + fillers.stream()
+                    .map(FillerDefinition::toString)
+                    .collect(Collectors.joining(", ")));
+
+    // Find all Caches that the opened inventory is part of
+    final List<Fillable> fillables = new ArrayList<>();
+    for (Cache cache : caches) {
+      if (filterPredicate.test(cache.region()) && filterPredicate.test(cache.filter())) {
```

Comment:
doesn't testing both separately sound alot like testing an `AllFilter.off(cache.region(), cache.filter())`, if that's the case it may be worth having the caches create one internally and using it here, but this is probably not needed if this is the only place we're checking this